components/style/values/specified/length.rs

```diff
@@ -40,18 +42,77 @@ impl ToCss for FontRelativeLength {
 }
 
 impl FontRelativeLength {
-    pub fn to_computed_value(&self,
-                             reference_font_size: Au,
-                             root_font_size: Au)
-                             -> Au
-    {
+    pub fn find_first_available_font_metrics(context: &Context) -> Option<FontMetrics> {
+        use font_metrics::FontMetricsQueryResult::*;
+        if let Some(ref metrics_provider) = context.font_metrics_provider {
+            for family in context.style().get_font().font_family_iter() {
+                if let Available(metrics) = metrics_provider.query(family.atom()) {
+                    return metrics;
+                }
+            }
+        }
+
+        None
+    }
+
+    // NB: The use_inherited flag is used to special-case the computation of
+    // font-family.
+    pub fn to_computed_value(&self, context: &Context, use_inherited: bool) -> Au {
+        let reference_font_size = if use_inherited {
+            context.inherited_style().get_font().clone_font_size()
+        } else {
+            context.style().get_font().clone_font_size()
+        };
+
+        let root_font_size = context.style().root_font_size;
         match *self {
             FontRelativeLength::Em(length) => reference_font_size.scale_by(length),
-            FontRelativeLength::Ex(length) | FontRelativeLength::Ch(length) => {
-                // https://github.com/servo/servo/issues/7462
-                let em_factor = 0.5;
-                reference_font_size.scale_by(length * em_factor)
+            FontRelativeLength::Ex(length) => {
+                match Self::find_first_available_font_metrics(context) {
+                    Some(metrics) => metrics.x_height,
+                    // https://drafts.csswg.org/css-values/#ex
+                    //
+                    //     In the cases where it is impossible or impractical to
+                    //     determine the x-height, a value of 0.5em must be
+                    //     assumed.
+                    //
+                    None => reference_font_size.scale_by(0.5 * length),
+                }
             },
+            FontRelativeLength::Ch(length) => {
+                let wm = context.style().writing_mode;
+
+                // FIXME(emilio): text-orientation: upright is only in Gecko
+                // right now.
+                let vertical = wm.is_vertical() || wm.is_vertical_lr();
```

Comment:
I think `sideways` values for both writing-mode and text-orientation will count as not-vertical here, since the text itself is in line with the inline flow. Not sure. We don't really support this yet (https://github.com/servo/servo/issues/14144)